### PR TITLE
Add Semantic Search to Public API

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds semantic search to the public API of `azure-search-documents`.

### CODE REVIEW

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/converters/SearchResultHelper.java [hunk 1]
```diff
@@ -47,4 +50,12 @@ static void setHighlights(SearchResult searchResult, Map<String, List<String>> h
     static void setJsonSerializer(SearchResult searchResult, JsonSerializer jsonSerializer) {
         accessor.setJsonSerializer(searchResult, jsonSerializer);
     }
+
+    static void setRerankerScore(SearchResult searchResult, Double rerankerScore) {
+        accessor.setRerankerScore(searchResult, rerankerScore);
+    }
+
+    static void setCaptions(SearchResult searchResult, List<CaptionResult> captions) {
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/converters/SearchResultHelper.java:58]: Question on the API: Does the order matters for the List<CaptionResult>?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchDocumentsResult.java [hunk 2]
```diff
@@ -46,7 +47,7 @@
      * query parameter was not specified or set to 'none'.
      */
     @JsonProperty(value = "@search.answers", access = JsonProperty.Access.WRITE_ONLY)
-    private Map<String, List<AnswerResult>> answers;
+    private List<AnswerResult> answers;
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchDocumentsResult.java:50]: Is this intended?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java [hunk 3]
```diff
@@ -370,8 +373,8 @@ public SearchOptions setMinimumCoverage(Double minimumCoverage) {
      * @param orderBy the orderBy value to set.
      * @return the SearchOptions object itself.
      */
-    public SearchOptions setOrderBy(List<String> orderBy) {
-        this.orderBy = orderBy;
+    public SearchOptions setOrderBy(String... orderBy) {
+        this.orderBy = (orderBy == null) ? null : java.util.Arrays.asList(orderBy);
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java:377]: nit:  `java.util.Arrays.asList` could be imported

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+     * @param name a name to look for.
+     * @return the corresponding QueryAnswer.
+     */
+    @JsonCreator
+    public static QueryAnswer fromString(String name) {
+        return fromString(name, QueryAnswer.class);
+    }
+
+    /** @return known QueryAnswer values. */
+    public static Collection<QueryAnswer> values() {
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java:33]: do we always need to have values() for the expandable string enum?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java [hunk 5]
```diff
@@ -19,42 +19,42 @@
...
+    @JsonProperty(value = "includeTotalCount")
     private Boolean includeTotalCount;
 
     /*
      * The list of facet expressions to apply to the search query. Each facet
      * expression contains a field name, optionally followed by a
      * comma-separated list of name:value pairs.
      */
-    @JsonProperty(value = "")
+    @JsonProperty(value = "Facets")
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:30]: is this shoud be lower-case for the  first letter

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java [hunk 6]
```diff
@@ -19,42 +19,42 @@
...
     private List<String> facets;
 
     /*
      * The OData $filter expression to apply to the search query.
      */
-    @JsonProperty(value = "")
+    @JsonProperty(value = "$filter")
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:36]: $filter? is this correct?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java [hunk 7]
```diff
@@ -19,42 +19,42 @@
...
     private String filter;
 
     /*
      * The list of field names to use for hit highlights. Only searchable
      * fields can be used for hit highlighting.
      */
-    @JsonProperty(value = "")
+    @JsonProperty(value = "HighlightFields")
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:43]: same here, lower case?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java [hunk 8]
```diff
@@ -95,14 +94,14 @@
      * called 'mylocation' the parameter string would be
      * "mylocation--122.2,44.8" (without the quotes).
      */
-    @JsonProperty(value = "")
+    @JsonProperty(value = "ScoringParameters")
```
Comment (mssfang) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:97]: this file has other places that have using the capital first letter.

**Review by mssfang — APPROVED** [review]

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/converters/SearchResultHelper.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/converters/SearchResultHelper.java:58]: Order doesn't matter

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchDocumentsResult.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchDocumentsResult.java:50]: Yes, the Swagger was slightly incorrect and the proper wire type was array.

https://github.com/Azure/azure-rest-api-specs/issues/13848

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java:377]: This is a code customization change where auto-import isn't able to easily resolve this, used a fully-qualified call while more functionality is added to the customization framework.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java:33]: @jianghaolu, @srnagar, is this a more recent codegen feature?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:30]: Hmm, good catch on this, the JSON name is `facets` so I'm not certain why this is being generated with capitalization unless this property in copied over to another value which is using the correct JSON naming. I'll follow-up on the Swagger for this.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java (hunk 4 above)
Comment (srnagar) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/QueryAnswer.java:33]: Nope, `values()` method is always generated for ExpandableStringEnum types.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/SearchOptions.java:30]: These values are passed into another method where these annotations aren't applied.

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java [hunk 9]
```diff
@@ -192,26 +194,26 @@
...
+     * is false. Setting this value to true may have a performance impact. Note that the count returned is an
      * approximation.
      *
-     * @param includeTotalResultCount the includeTotalResultCount value to set.
+     * @param includeTotalCount the includeTotalCount value to set.
      * @return the SearchOptions object itself.
      */
-    public SearchOptions setIncludeTotalResultCount(Boolean includeTotalResultCount) {
-        this.includeTotalResultCount = includeTotalResultCount;
+    public SearchOptions setIncludeTotalCount(Boolean includeTotalCount) {
```
Comment (srnagar) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java:215]: the names of getter and setter don't match. Can this be updated to `setTotalCountIncluded`?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/AutocompleteOptions.java [hunk 10]
```diff
@@ -1,34 +1,31 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+// Code generated by Microsoft (R) AutoRest Code Generator.
+// Changes may cause incorrect behavior and will be lost if the code is
+// regenerated.
```
Comment (srnagar) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/AutocompleteOptions.java:5]: Was this comment manually added?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedFlux.java [hunk 11]
```diff
@@ -101,4 +102,20 @@ public SearchPagedFlux(Supplier<Mono<SearchPagedResponse>> firstPageRetriever,
...
+     * The answer results based on the search request.
+     * <p>
+     * If {@code answers} wasn't supplied in the request this will be null.
+     *
+     * @return The answer results if {@code answers} were supplied in the request, otherwise null.
+     */
+    public Mono<List<AnswerResult>> getAnswers() {
+        return metadataSupplier.get()
+            .flatMap(metaData -> {
+                if (metaData.getFirstPageResponse().getAnswers() == null) {
```
Comment (srnagar) [inline sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedFlux.java:115]: Will all the results of `getAnswers()` always be on the first page? Also, if the user has used `byPage()` or simply subscribed to the paged flux, is it necessary to make another call to the service? Can we cache the answers from the first call that happened through other APIs?

File: sdk/search/azure-search-documents/swagger/src/main/java/SearchIndexCustomizations.java [hunk 12]
```diff
@@ -1,11 +1,88 @@
...
+//                "this.scoringParameters = (scoringParameters == null)"
+//                    + "    ? null"
+//                    + "    : java.util.Arrays.stream(scoringParameters)"
+//                    + "        .map(ScoringParameter::toString)"
+//                    + "        .collect(java.util.stream.Collectors.toList());"
+//                    + "return this;");
+//
+//        classCustomization.getMethod("getScoringParameters")
+//            .setReturnType("List<ScoringParameter>",
+//                "this.scoringParameters.stream().map(ScoringParameter::new).collect(java.util.stream.Collectors.toList())");
```
Comment (srnagar) [inline sdk/search/azure-search-documents/swagger/src/main/java/SearchIndexCustomizations.java:86]: Can this be removed and added later when this is supported?

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/AutocompleteOptions.java (hunk 10 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/AutocompleteOptions.java:5]: Nope, I've transitioned this class over to being code generated with customizations

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java (hunk 9 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/models/SearchOptions.java:215]: The reason for the mismatch is I'm trying to get this more inline with the hand-written class to hopefully replace that soon with this generated class

File: sdk/search/azure-search-documents/swagger/src/main/java/SearchIndexCustomizations.java (hunk 12 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/swagger/src/main/java/SearchIndexCustomizations.java:86]: I'd rather leave it as a reminder of what needs to change to enable the code generated class matching what the hand-written class provides

File: sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedFlux.java (hunk 11 above)
Comment (alzimmermsft) [reply sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedFlux.java:115]: I think this is a good idea to make it so another API call isn't being made, mind filling an issue for this as this is pre-existing.

**Pull request merged**